File: 10205/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int set = 0;
int deck[2][52];
int shuffles[100][52];

char *names[] = { "Ace", "2", "3", "4", "5", "6", "7", "8", "9", "10", "Jack", "Queen", "King"};
char *color[] = { "Clubs", "Diamonds", "Hearts", "Spades"};

void shuffle(int n){
    int i;
    for(i = 0; i < 52; i++)
        deck[1 - set][i] = deck[set][ shuffles[n][i] - 1 ];
    set = 1 - set;
}


void get_shuffles(){
    int i, j, n;
    scanf("%d", &n);
    for(i = 0; i < n; i++)
        for(j = 0; j < 52; j++)
            scanf("%d\n", &shuffles[i][j]);
}


void solve(){
    char s[5];
    while(gets(s) && strlen(s))
        shuffle(atoi(s) - 1);
}


void init(){
    int i;
    for(i = 0; i < 52; i++)
        deck[0][i] = deck[1][i] = i + 1;
}


void print(){
    int i;
    for(i = 0; i < 52; i++)
        printf("%s of %s\n", names[ (deck[set][i]) % 13 ], color[ (deck[set][i]-1) / 13 ]);
}


int main()
{
    int N;
    scanf("%d", &N);
    while(N--){
        init();
        get_shuffles();
        solve();
        print();
        if(N > 0) puts("");
    }
    return 0;
}

File: 10205/Main.java
import java.util.Scanner;

public class Main {
    static int Set = 0;
    static int[][] Deck = new int[2][52];
    static int[][] Shuffles = new int[100][52];
    static String[] Names = new String[]{ "Ace", "2", "3", "4", "5", "6", "7", "8", "9", "10", "Jack", "Queen", "King"};
    static String[] Color = new String[]{ "Clubs", "Diamonds", "Hearts", "Spades"};
    static Scanner sc  = new Scanner(System.in);
    public static void main(String[] args) {
        int N;
        N = Integer.parseInt(sc.nextLine());
        sc.nextLine();
        while(N-- != 0){
            init();
            get_shuffles();
            solve();
            print();
            if(N > 0)
                System.out.println();
        }
    }

    private static void print() {
        int i;
        for(i = 0; i < 52; i++)
            System.out.println(Names[ (Deck[Set][i]) % 13] + " of " + Color[ (Deck[Set][i]-1) / 13]);
    }

    private static void solve() {
        String numShuffle;
            while (sc.hasNextLine() && !(numShuffle = sc.nextLine()).equals("")) {
                shuffle(Integer.parseInt(numShuffle) - 1);
            }
    }

    private static void shuffle(int n){
        int i;
        for(i = 0; i < 52; i++) {
            if(Shuffles[n][i] == 0)
                System.out.println(n + " " + i);
            Deck[1 - Set][i] = Deck[Set][Shuffles[n][i] - 1];
        }
        Set = 1 - Set;
    }

    private static void get_shuffles() {
        int i, j, n, cardCounter;
        n = Integer.parseInt(sc.nextLine());
        for(i = 0; i < n; i++) {
            cardCounter = 0;
            while(cardCounter != 52) {
                String[] x = sc.nextLine().split(" ");
                for (j = cardCounter; j < cardCounter + x.length; j++) {
                    Shuffles[i][j] = Integer.parseInt(x[j - cardCounter]);
                }
                cardCounter += x.length;
            }
        }
    }

    private static void init() {
        int i;
        for(i = 0; i < 52; i++)
            Deck[0][i] = Deck[1][i] = i + 1;
    }
}
